lib/sfu/sfu.ex

```diff
@@ -0,0 +1,396 @@
+    end
+  end
+
+  defp handle_media_event(%{type: :sdp_answer} = event, peer_id, ctx, state) do
+    actions = [
+      forward: {{:endpoint, peer_id}, {:signal, {:sdp_answer, event.data.sdp_answer.sdp}}}
+    ]
+
+    {tracks_msgs, state} =
+      if Map.has_key?(state.incoming_peers, peer_id) do
+        inbound_tracks = Map.values(state.endpoints[peer_id].inbound_tracks)
+        {peer, state} = pop_in(state, [:incoming_peers, peer_id])
+        peer = Map.delete(peer, :tracks_metadata)
+        peer = Map.put(peer, :mid_to_track_metadata, event.data.mid_to_track_metadata)
+        state = put_in(state, [:peers, peer_id], peer)
+        tracks_msgs = update_track_messages(ctx, inbound_tracks, {:endpoint, peer_id})
+
+        MediaEvent.create_peer_joined_event(
+          peer_id,
+          state.peers[peer_id].metadata,
+          event.data.mid_to_track_metadata
+        )
+        |> dispatch(state)
+
+        {tracks_msgs, state}
+      else
+        {[], state}
+      end
+
+    {actions ++ tracks_msgs, state}
+  end
+
+  defp handle_media_event(%{type: :candidate} = event, peer_id, _ctx, state) do
+    actions = [forward: {{:endpoint, peer_id}, {:signal, {:candidate, event.data.candidate}}}]
+    {actions, state}
+  end
+
+  defp handle_media_event(%{type: :leave}, peer_id, ctx, state) do
+    {actions, state} = remove_peer(peer_id, ctx, state)
+    {actions, state}
+  end
+
+  @impl true
+  def handle_notification({:signal, message}, {:endpoint, peer_id}, _ctx, state) do
+    MediaEvent.create_signal_event(peer_id, {:signal, message})
+    |> dispatch(state)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_notification({:new_track, track_id, encoding}, endpoint_bin_name, ctx, state) do
+    Membrane.Logger.info(
+      "New incoming #{encoding} track #{track_id} from #{inspect(endpoint_bin_name)}"
+    )
+
+    {:endpoint, endpoint_id} = endpoint_bin_name
+
+    tee = {:tee, {endpoint_id, track_id}}
+    fake = {:fake, {endpoint_id, track_id}}
+
+    children = %{
+      tee => Membrane.Element.Tee.Master,
+      fake => Membrane.Element.Fake.Sink.Buffers
+    }
+
+    extensions = setup_extensions(encoding, state[:options][:extension_options])
+
+    links =
+      [
+        link(endpoint_bin_name)
+        |> via_out(Pad.ref(:output, track_id), options: [extensions: extensions])
+        |> to(tee)
+        |> via_out(:master)
+        |> to(fake)
+      ] ++
+        flat_map_children(ctx, fn
+          {:endpoint, other_endpoint_id} = other_endpoint_name ->
+            if endpoint_bin_name != other_endpoint_name and
+                 state.endpoints[other_endpoint_id].ctx.receive_media do
+              [
+                link(tee)
+                |> via_out(:copy)
+                |> via_in(Pad.ref(:input, track_id), options: [encoding: encoding])
+                |> to(other_endpoint_name)
+              ]
+            else
+              []
+            end
+
+          _child ->
+            []
+        end)
```

Comment:
```suggestion
          {:endpoint, other_endpoint_id} = other_endpoint_name
            when endpoint_bin_name != other_endpoint_name
            and state.endpoints[other_endpoint_id].ctx.receive_media ->
            [
              link(tee)
              |> via_out(:copy)
              |> via_in(Pad.ref(:input, track_id), options: [encoding: encoding])
              |> to(other_endpoint_name)
            ]
          _child ->
            []
        end)
```

Reply:
Cannot do `state.endpoints[other_endpoint_id]` in guard :/

Reply:
hmm, so you may consider `:erlang.map_get(other_endpoint_id, state.endpoints)